src/Chainweb/Utils.hs

```diff
@@ -407,6 +408,11 @@ minusOrZero :: Ord a => Num a => a -> a -> a
 minusOrZero a b = a - min a b
 {-# INLINE minusOrZero #-}
 
+-- | Analogous to `unsafeInterleaveIO` but doesn't hide the effect behind evaluation.
+interleaveIO :: IO a -> IO (IO a)
+interleaveIO act = evaluate <$> unsafeInterleaveIO act
+{-# INLINE interleaveIO #-}
```

Comment:
This is an interesting implementation of [`once`](https://hackage.haskell.org/package/io-memoize-1.1.1.0/docs/System-IO-Memoize.html#:~:text=once%20%3A%3A%20IO%20a%20%2D%3E%20IO%20(IO%20a)), (see also this [`once`](https://hackage.haskell.org/package/once-0.4/docs/Control-Once.html#v:once)) without introducing any dependency. So I'd recommend renaming this function to `once` since that's what it's often called.

Once caveat of this implementation is what happens in the presence of an async exception:
``` haskell
ghci> d <- evaluate <$> unsafeInterleaveIO (threadDelay 10000000)
ghci> d
^CInterrupted.
ghci> d
Interrupted.
```

I.e. the async exception gets trapped inside `a`, so I'd say this is a semantically slippery ground since an arbitrary `timeout` somewhere in the code has a chance of being the first one to ask for `a` and poisoning it for everyone else.

In comparison, [`io-memoize`](https://hackage.haskell.org/package/io-memoize-1.1.1.0/docs/System-IO-Memoize.html) retries the action in that case:
``` haskell
ghci> d <- once $ putStrLn "Starting" >> threadDelay 10000000
ghci> d
Starting
^CInterrupted.
ghci> d
Starting
```
Though to be fair, `io-memoize` always retries if it encounters an exception:
``` haskell
ghci> d <- once $ putStrLn "Starting" >> threadDelay 10000000 >> error "FAIL!"
ghci> d
Starting
*** Exception: FAIL!
CallStack (from HasCallStack):
  error, called at <interactive>:9:60 in interactive:Ghci5
ghci> d
Starting
*** Exception: FAIL!
CallStack (from HasCallStack):
  error, called at <interactive>:9:60 in interactive:Ghci5
```

I don't think there's any "right" behavior here, retrying on async exceptions might also invalidate the caller's assumption that the IO inside can be executed at most once, so I guess the least we can do is to mention this behavior in the haddock comment.

Reply:
This is not the behavior that I expected, so I'll leave a comment and rename.

Reply:
+1 to renaming and commenting about the behaviour.